x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/custom/request/CustomRequest.java

```diff
@@ -0,0 +1,165 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.services.custom.request;
+
+import org.apache.http.HttpHeaders;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.entity.StringEntity;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.inference.TaskType;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.inference.common.ValidatingSubstitutor;
+import org.elasticsearch.xpack.inference.external.request.HttpRequest;
+import org.elasticsearch.xpack.inference.external.request.Request;
+import org.elasticsearch.xpack.inference.services.custom.CustomModel;
+import org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings;
+import org.elasticsearch.xpack.inference.services.settings.SerializableSecureString;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.inference.common.JsonUtils.toJson;
+import static org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings.REQUEST_CONTENT;
+import static org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings.URL;
+
+public class CustomRequest implements Request {
+    private static final String QUERY = "query";
+    private static final String INPUT = "input";
+
+    private final URI uri;
+    private final ValidatingSubstitutor jsonPlaceholderReplacer;
+    private final ValidatingSubstitutor stringPlaceholderReplacer;
+    private final CustomModel model;
+
+    public CustomRequest(String query, List<String> input, CustomModel model) {
+        this.model = Objects.requireNonNull(model);
+
+        var stringOnlyParams = new HashMap<String, String>();
+        addStringParams(stringOnlyParams, model.getSecretSettings().getSecretParameters());
+        addStringParams(stringOnlyParams, model.getTaskSettings().getParameters());
+
+        var jsonParams = new HashMap<String, String>();
+        addJsonStringParams(jsonParams, model.getSecretSettings().getSecretParameters());
+        addJsonStringParams(jsonParams, model.getTaskSettings().getParameters());
+
+        if (query != null) {
+            jsonParams.put(QUERY, toJson(query, QUERY));
+        }
+
+        addInputJsonParam(jsonParams, input, model.getTaskType());
+
+        jsonPlaceholderReplacer = new ValidatingSubstitutor(jsonParams, "${", "}");
+        stringPlaceholderReplacer = new ValidatingSubstitutor(stringOnlyParams, "${", "}");
+        uri = buildUri();
+    }
+
+    private static void addStringParams(Map<String, String> stringParams, Map<String, ?> paramsToAdd) {
```

Comment:
Fields like the url, query parameters, and headers should not have their values converted to json format. This only accepts strings and doesn't manipulate them.